src/Functions/fromReadableSize.cpp

```diff
@@ -0,0 +1,265 @@
+                        "Invalid expression for function {}, convert string to float64 failed: \"{}\".",
+                        getName(),
+                        String(base_str));
+                }
+                base = value.get();
+
+                scanSpaces(str, token_tail, last_pos);
+                token_front = token_tail;
+
+                /// scan a unit
+                if (!scanUnit(str, token_tail, last_pos))
+                {
+                    throw Exception(
+                        ErrorCodes::BAD_ARGUMENTS,
+                        "Invalid expression for function {}, find unit failed, str: \"{}\".",
+                        getName(),
+                        String(str));
+                }
+
+                std::string unit = std::string{str.substr(token_front, token_tail - token_front)};
+                boost::algorithm::to_lower(unit);
+                auto iter = size_unit_to_bytes.find(unit);
+                if (iter == size_unit_to_bytes.end()) /// not find unit
+                {
+                    throw Exception(
+                        ErrorCodes::BAD_ARGUMENTS, "Invalid expression for function {}, parse unit failed: \"{}\".", getName(), unit);
+                }
+                Float64 raw_num_bytes = base * iter->second;
+                if (raw_num_bytes > MAX_UINT64)
+                {
+                    throw Exception(
+                        ErrorCodes::BAD_ARGUMENTS,
+                        "Invalid expression for function {}, result is too big for output data type (UInt64): \"{}\".",
+                        getName(),
+                        raw_num_bytes
+                    );
+                }
+                // As the input might be an arbitrary decimal number we might end up with a non-integer amount of bytes when parsing binary (eg MiB) units.
+                // This doesn't make sense so we round up to indicate the byte size that can fit the passed size.
+                result = static_cast<UInt64>(std::ceil(raw_num_bytes));
+
+                res_data.emplace_back(result);
+            }
+
+            return col_to;
+        }
+
+        /// scan an unsigned integer number
+        static bool scanUnsignedInteger(std::string_view & str, Int64 & index, Int64 last_pos)
+        {
+            int64_t begin_index = index;
+            while (index <= last_pos && isdigit(str[index]))
+            {
+                index++;
+            }
+            return index != begin_index;
+        }
+
+        /// scan a unit
+        static bool scanUnit(std::string_view & str, Int64 & index, Int64 last_pos)
+        {
+            int64_t begin_index = index;
+            while (index <= last_pos && !isdigit(str[index]) && !isSeparator(str[index]))
+            {
+                index++;
+            }
+            return index != begin_index;
+        }
+
+        /// scan spaces
+        static void scanSpaces(std::string_view & str, Int64 & index, Int64 last_pos)
```

Comment:
Comments in l. 200, 211 and 222 are useless.

Should we rename this function `skipSpaces`?

Let's pass `str` by const reference.

Please no in-out parameters (`index`), if not necessary (which it isn't). Please return the new index by value.

Reply:
[The parser auxiliary methods have been removed](https://github.com/ClickHouse/ClickHouse/pull/64386#discussion_r1616168619) as they're not needed anymore, together with their pointless comments.